### DISCUSSION TITLE
How do I connect to Oracle?

### DISCUSSION THREAD

**Discussion by abel-diaz** [sfu-db/connector-x#130] [Q&A] [answered]
I am unable to connect to Oracle. I am using Connector X v. 0.2.1a1
I have the following creds
host
port
servicename
username
password

According to the connectorx instructions i need to have 
(postgres|postgressql|mysql|mssql|sqlite)://username:password@addr:port/dbname

Using my above creds, I have
cstr = 'oracle://username:password@host:port/servicename'
query ' 'select * from table'

When I run 
df = cx.read_sql(cstr,query)

I receive RuntimeError: timed out waiting for connection: OCI Error: ORA-01017: invalid username/password; logon denied
I am able to connect with sqlalchemy using my creds

How do I connect conncetorx with Oracle?
Thank you,
Abel

**Comment by wangxiaoying** [comment]
Hi @abel-diaz ,

I believe your connection string and the way you invoke `cx.read_sql` is correct.

Our connection string should look like: 

```bash
oracle://${ORACLE_USER}:${ORACLE_PASSWORD}@${ORACLE_HOST}:${ORACLE_PORT}/${ORACLE_DB}
```

, which is different from using `sqlplus`.

Do you have any special characters in your username or password? I'm wondering maybe the issue is in the url parsing and some strings need to be [quoted](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.quote_plus).

**Reply by abel-diaz** [reply]
Good Morning,
If my creds have the following ( I changed the words but kept special characters)
host - hostname.hostsite.com
port - 1234
servicename - dbprod
username - myname[text]
password - zoo^123

I am currently writing the connection as 
cstr = 'oracle://myname[text]:zoo^1234@hostname.hostsite.com:1234/dbprod'

How would I correct the above connection string to make it work? 
Thank you for your help
Abel

**Reply by wangxiaoying** [reply]
Can you try to encode your username and password like this and see whether it works?

```Python
import urllib.parse

cstr = f"oracle://{urllib.parse.quote_plus('myname[text]')}:{urllib.parse.quote_plus('zoo^1234')}@hostname.hostsite.com:1234/dbprod"
```

**Comment by abel-diaz** [comment, marked as answer]
Hi,
I tried (with actual creds) 
cstr = f"oracle://{urllib.parse.quote_plus('myname[text]')}:{urllib.parse.quote_plus('zoo^1234')}@hostname.hostsite.com:1234/dbprod"

and the output for cstr is similar to below with example creds
'oracle://myname%5Btext%5D:zoo%5E1234@hostname.hostsite.com:1234/dbprod'

When i try 
df = cx.read_sql(cstr,query)

I am met with 
Error: ORA-01017: invalid username/password; logon denied

I am really hoping this gets connected. It takes me 1hr and 40min to pull 3M rows of SQL to DF using pd.read_sql. 

Thanks for all the help. I have confirmed all my creds are correct through my companies DBA team and thru successful connections in TOAD, cx_oracle, and sqlAlchemy. 

Thank you again
Abel

**Reply by dovahcrow** [reply]
Hi @abel-diaz , can you confirm `myname%5Btext%5D:zoo%5E1234` is correct? it seems your password is URL encoded.

**Reply by abel-diaz** [reply]
Hi
Yes my username and password are entered correctly
username - myname[text]
password - zoo^123

I have a username structured just as above with brackets but with different text
I have a password structured just as above with text^numbers

**Reply by wangxiaoying** [reply]
Hi @abel-diaz , I tried to create two users in oracle:

* user1: name=myname[text], password=zoo^123
* user2: name=myname, password=zoo^123

and use `cx_oracle` driver through `sqlalchemy` to connect to oracle, and I found user1 fails to connect and user2 succeed.

I also tried using our library and we failed to connect for both users. The reason is we automatically do urlencoding on the password field, which oracle cannot recognize. I did a quick fix in our library and can successfully connect to oracle using user2's username and password now. But still user1 cannot connect to oracle. I tried both `sqlplus` client and `cx_oracle` driver through `sqlalchemy` and they failed to connect as well. (The same error: `invalid username/password; logon denied`)

If it is possible that you could use a username without special character (special character in password seems fine)? If that's possible, I can release an alpha version for you, which fix the above issue that I mentioned.

**Reply by abel-diaz** [reply]
Hi,
this is great news for all of us oracle users! Yes I verified I can use username 2 "myname" , username 1 myname[text] is a proxy connect. Both should connect, as told by my DBA team. Please release an alpha version, version name, and I can pip install it. I am very excited to use connector x.  This sounds promising!

**Reply by wangxiaoying** [reply]
Hi, we just released the new alpha version (install by: `pip install -U connectorx==0.2.1a4`), please feel free to try it!

Also we are still trying to optimize the performance of Oracle (#127 ), so it is recommended to use partitions in order to leverage multiple cores when loading the data for now like this:

```Python
import connectorx as cx

partition_on = "column_name" # an integer column in result served as partition key (better be a key with index since we will evenly partition this column in each subquery)
partition_num = 10 # number of cores you want to use

df = cx.read_sql('oracle://myname:zoo^123@your_hostname:1521/testdb', 'select * from table', partition_on=partition_on, partition_num=partition_num)
```

Without any partitioning, the time might be only a little bit faster than pandas right now in our test, but the memory usage is still 3x less.

**Reply by abel-diaz** [reply]
Hi,
It was a success! username2 is my direct connect name, username1[schemaname] is a proxy connect. When I connected with username2 I got connected but could not access my data. I then changed my username to username1[schemaname] so my connection string looked as followed:
df = cx.read_sql('oracle://username1[schemaname]:zoo^123@your_hostname:1521/testdb', 'select * from table')

Using the new version (0.2.1a4) I was successfully connected and able to query. It handled the " [schemaname] " portion. 
Thank you so much for all your support and helping me (and im sure others) get connected. 

Now, its time to use it!
Thank you!
Abel

**Reply by abel-diaz** [reply]
PS - What is the max number of partitions I can assign? ... safely?

Thanks again

**Reply by wangxiaoying** [reply]
That's great! Also, thanks for reporting this bug!

> PS - What is the max number of partitions I can assign? ... safely?

It is recommended to be the number of free cores you have on your client machine.

**Reply by abel-diaz** [reply]
I have 4 cores and 8 logical processors, my max would be 4 , do logical processors play into this? If so, how 
I have hyper-threading so 4 cores, acts as 8 . 

Questions
1. Are you recommending to not use 8 as a max but stick to the true count of cores that is 4? 
2. Can connector x handle hyper threading? If so, can I set a max from 4 to 8?
3. Does the same apply to virtual processors in a VM?
4. How would I know if I over extended my max core allotment to connector x? Will it fail, what could I look out for to know?

Thanks
Abel

**Reply by wangxiaoying** [reply]
Oh, sorry I misunderstood your question before. I believe you could assign values even larger than 8, but the performance might not improved or even introduce overhead when you make it too large, so I suggested to use the number of cores you have. But in a specific case the optimal value might also be affected by other factors like the network condition and database server's performance.

**Reply by abel-diaz** [reply]
Ok thank you for your help!
